Shutdown children synchronously, in reverse-startup order

QuinnWilton opened the pull request:
This is a work in progress, and still needs some tests, but implements the behaviour we chatted about.

I opted to create a new field for `startup_index` rather than putting that information in the `meta` field, because it seems like the `meta` field is intended to be set by consumers of the library. If I'm misunderstanding things and you think there's a better place to track that, let me know and I'm happy to adjust it :)

When making shutdown synchronous I also changed it so that we shutdown timers only for the process that is being terminated. I believe this is more in-line with how `supervisor` works, but I need to verify that still.

Since we also immediately handle processes that need to be brutal killed, I removed the extra handling for those processes in `shutdown_all`.

Based on the existing test coverage things look like they're still working as intended, but I need to write some new tests to cover shutdown ordering.

Thanks for the help, and let me know if you disagree with anything in here :)

QuinnWilton committed:
Shutdown children in reverse startup order

QuinnWilton committed:
Shutdown children synchronously

Previously children would be simultaneously shutdown.
This changes the behaviour to be in line with how
supervisor works, by shutting down the children
one-by-one.

sasa1977 commented on lib/parent/functional.ex:
```diff
@@ -180,50 +181,43 @@ defmodule Parent.Functional do
-
-    # Kill all timeout timers first, because we don't want timeout to interfere with the shutdown logic.
-    Enum.each(pids_and_specs, fn {pid, process} -> kill_timer(process.data.timer_ref, pid) end)
+      |> Enum.sort_by(fn {_pid, process} -> process.data.startup_index end, :desc)
```
Since we're using the new sort, we should add Elixir 1.10 as a minimum version in mix.exs.

QuinnWilton replied:
Good catch! To avoid bumping the Elixir version would you prefer if I didn't use the new API?

sasa1977 replied:
I personally don't try too hard to keep compatibility with older Elixir versions, because that usually adds more complexity, and also because upgrading to the most recent Elixir is IME typically not very complicated. That said, making this work on pre 1.10 is trivial, so I'd slightly prefer that.

I wonder why CI failure isn't shown here. I can see the failure on the travis site, but don't see anything here. I probably need to configure something. I'll see if I can make that work.

sasa1977 reviewed (commented):
Thanks, it looks great so far!

I looked at the tests briefly, and I feel that I overengineereed them when I originally wrote the code. I think the simplest way to test this is to write an example based test (so `test` not a `property`) in functional_test.exs. Basically start some children, collect their pids, monitor each started child, stop all children, collect all `:DOWN` messages (without pattern matching on the exact pid), and get the pids. Finally, assert that the pids are collected in the reverse order. Perhaps there's a better way, but this is what I came up with after thinking about the problem for about 1 minute :-)

I don't think that tests in `Procdict` and `GenServer` are required. TBH if I was writing this code from scratch today, I'd probably move the `Procdict` logic into the `GenServer` module (`Functional` and `Registry` would remain), and test only at the `GenServer` level, using example-based tests. I believe that such code would be simpler to work with. I'll think about refactoring this code going forward, but that's just a digression.

QuinnWilton commented:
I've pushed up changes addressing your comments!

I modified the call to `Enum.sort_by/3` to simply pass `&>=/2` as a sorter, to avoid requiring an Elixir version bump, and I added tests for the termination order. Since the property tests already provided most of the machinery needed, I opted to just add a little bit of extra logic to the existing `"shutting down all children"` property. If you'd prefer a more explicit example test I'm happy to add that too :)

One of the builds is failing on Travis, but I haven't managed to reproduce locally yet. I'll update the PR once I track down the issue: https://travis-ci.org/github/sasa1977/parent/jobs/672219756

QuinnWilton commented:
I made one other minor change: https://github.com/sasa1977/parent/pull/9/files#diff-497ad4ae272b61f81bd8910b85c3f949R226

If a process failed to terminate within the the allotted time, we would try to kill it. While waiting for the process to be killed, however, we would use the original shutdown timeout for the process, rather than `:infinity`. This means that if the process failed to terminate in time (unlikely in practice, but I can imagine it happening if the schedulers are at 100% utilization) we would hit the timeout and then try to kill the process again.

I don't think this had any negative impact, but it's a little bit of a confusing flow, so I made sure to use a timeout of `:infinity` when we are killing the process.

sasa1977 commented on test/parent/functional_test.exs:
```diff
@@ -199,4 +206,13 @@ defmodule Parent.FunctionalTest do
     end
   end
+
+  defp await_monitored_processes(0), do: []
+
+  defp await_monitored_processes(n) do
+    receive do
+      {:DOWN, _ref, :process, pid, _reason} ->
+        [pid | await_monitored_processes(n - 1)]
+    end
+  end
```
Alternatively without recursion, and using some ExUnit helpers (untested):

```elixir
Stream.repeatedly(fn -> 
  assert_received({:DOWN, _ref, :process, pid, _reason})
  pid
end)
|> Enum.take(n)
```

sasa1977 reviewed (approved):
> While waiting for the process to be killed, however, we would use the original shutdown timeout for the process, rather than `:infinity`.

Great catch! Yeah, I agree with you.



Thanks, this looks great! I'll merge this now and do some post-merge cleanup (like changelog & docs) these days, and then release the next version. Thank you for contributing!

QuinnWilton commented:
Thank you for so quickly reviewing and talking through this issue with me, I really enjoyed working on it!

Don't be afraid to reach out if there's anything else you'd like help with.

sasa1977 commented:
FYI, these changes are now published to hex.
